realm/realm-library/src/main/java/io/realm/RealmResults.java

```diff
@@ -723,54 +727,54 @@ public boolean addAll(Collection<? extends E> collection) {
         int pos = -1;
 
         RealmResultsIterator() {
-            tableViewVersion = table.sync();
+            tableViewVersion = currentTableViewVersion;
         }
 
+        /**
+         * {@inheritDoc}
+         */
         public boolean hasNext() {
-            assertRealmIsStable();
             return pos + 1 < size();
         }
 
+        /**
+         * {@inheritDoc}
+         */
         public E next() {
-            assertRealmIsStable();
+            realm.checkIfValid();
+            checkRealmIsStable();
             pos++;
             if (pos >= size()) {
-                throw new IndexOutOfBoundsException("Cannot access index " + pos + " when size is " + size() +  ". Remember to check hasNext() before using next().");
+                throw new NoSuchElementException("Cannot access index " + pos + " when size is " + size() +  ". Remember to check hasNext() before using next().");
             }
             return get(pos);
         }
 
         /**
-         * Not supported by RealmResults.
+         * Not supported by RealmResults iterators.
          *
-         * @throws UnsupportedOperationException always.
+         * @throws {@link UnsupportedOperationException} always
          */
+        @Deprecated
         public void remove() {
-            throw new RealmException("Removing is not supported.");
-    /*        assertRealmIsStable();
-            if (pos == -1) {
-                throw new IllegalStateException("Must call next() before calling remove()");
-            }
-            if (removeUsed) {
-                throw new IllegalStateException("Cannot call remove() twice. Must call next() in between");
-            }
-
-            RealmResults.this.remove(pos);
-            pos--;
-            removeUsed = true;
-            currentTableViewVersion = getTable().sync();
-     */   }
+            throw new UnsupportedOperationException("remove() is not supported by RealmResults iterators.");
+        }
 
-        protected void assertRealmIsStable() {
-            long version = table.sync();
-            if (tableViewVersion > -1 && version != tableViewVersion) {
-                throw new ConcurrentModificationException("No outside changes to a Realm is allowed while iterating a RealmResults. Use iterators methods instead.");
+        protected void checkRealmIsStable() {
+            long version = table.getVersion();
+            // All changes inside a write transaction will continuously update the table version, and we can
```

Comment:
"Any changes within a write transaction will immediately update the table version"